File: app/src/main/java/com/cristhian/gametobeat/database/GameRepository.java
package com.cristhian.gametobeat.database;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;

import java.util.List;

public class GameRepository {

    private GameDao mGameDao;
    private LiveData<List<Game>> mGamesBeat;
    private LiveData<List<Game>> mGamesPlaying;
    private LiveData<List<Game>> mGamesWant;

    GameRepository(Application application) {
        GameBeatDataBase db = GameBeatDataBase.getInstance(application);
        mGameDao = db.gameDao();
        mGamesBeat = mGameDao.loadGamesByType(Game.STATUS_BEAT);
        mGamesPlaying = mGameDao.loadGamesByType(Game.STATUS_PLAYING);
        mGamesWant = mGameDao.loadGamesByType(Game.STATUS_WANT);
    }

    public LiveData<List<Game>> getGamesPlaying() {
        return mGamesPlaying;
    }

    public LiveData<List<Game>> getGamesBeat() {
        return mGamesBeat;
    }

    public LiveData<List<Game>> getGamesWant() {
        return mGamesWant;
    }

    public void insert(Game game) {
        new insertAsyncTask(mGameDao).execute(game);
    }

    public void delete(Game game) { new deleteAsyncTask(mGameDao).execute(game); }

    public void update(Game game) {
        new updateAsyncTask(mGameDao).execute(game);
    }

    public LiveData<Game> loadGameById(Game game) {
        return mGameDao.loadGameById(game.getGameId());
    }

    private static class insertAsyncTask extends AsyncTask<Game, Void, Void> {

        private GameDao mAsyncTaskDao;

        insertAsyncTask(GameDao gameDao) {
            mAsyncTaskDao = gameDao;
        }

        @Override
        protected Void doInBackground(Game... games) {
            mAsyncTaskDao.insertGame(games[0]);
            return null;
        }
    }

    private static class deleteAsyncTask extends AsyncTask<Game, Void, Void> {

        private GameDao mAsyncTaskDao;

        deleteAsyncTask(GameDao gameDao) {
            mAsyncTaskDao = gameDao;
        }

        @Override
        protected Void doInBackground(Game... games) {
            mAsyncTaskDao.deleteGame(games[0]);
            return null;
        }
    }

    private static class updateAsyncTask extends AsyncTask<Game, Void, Void> {

        private GameDao mAsyncTaskDao;

        updateAsyncTask(GameDao gameDao) {
            mAsyncTaskDao = gameDao;
        }

        @Override
        protected Void doInBackground(Game... games) {
            mAsyncTaskDao.updateGame(games[0]);
            return null;
        }
    }
}

File: app/src/main/java/com/cristhian/gametobeat/database/GameRepository.kt
package com.cristhian.gametobeat.database

import android.app.Application
import android.os.AsyncTask
import androidx.lifecycle.LiveData
import com.cristhian.gametobeat.database.GameBeatDataBase.Companion.getInstance

class GameRepository internal constructor(application: Application?) {
    private val mGameDao: GameDao
    val gamesBeat: LiveData<List<Game?>?>?
    val gamesPlaying: LiveData<List<Game?>?>?
    val gamesWant: LiveData<List<Game?>?>?

    fun insert(game: Game?) {
        insertAsyncTask(mGameDao).execute(game)
    }

    fun delete(game: Game?) {
        deleteAsyncTask(mGameDao).execute(game)
    }

    fun update(game: Game?) {
        updateAsyncTask(mGameDao).execute(game)
    }

    fun loadGameById(game: Game): LiveData<Game?>? {
        return mGameDao.loadGameById(game.gameId)
    }

    private class insertAsyncTask internal constructor(private val mAsyncTaskDao: GameDao) :
            AsyncTask<Game?, Void?, Void?>() {
        override fun doInBackground(vararg games: Game?): Void? {
            games[0]?.let { mAsyncTaskDao.insertGame(it) }
            return null
        }

    }

    private class deleteAsyncTask internal constructor(private val mAsyncTaskDao: GameDao) :
            AsyncTask<Game?, Void?, Void?>() {
        override fun doInBackground(vararg games: Game?): Void? {
            games[0]?.let { mAsyncTaskDao.deleteGame(it) }
            return null
        }

    }

    private class updateAsyncTask internal constructor(private val mAsyncTaskDao: GameDao) :
            AsyncTask<Game?, Void?, Void?>() {
        override fun doInBackground(vararg games: Game?): Void? {
            games[0]?.let { mAsyncTaskDao.updateGame(it) }
            return null
        }

    }

    init {
        val db = getInstance(application!!)
        mGameDao = db!!.gameDao()
        gamesBeat = mGameDao.loadGamesByType(Game.STATUS_BEAT)
        gamesPlaying = mGameDao.loadGamesByType(Game.STATUS_PLAYING)
        gamesWant = mGameDao.loadGamesByType(Game.STATUS_WANT)
    }
}